src/github.com/matrix-org/dendrite/clientapi/readers/profile.go

```diff
@@ -175,13 +258,68 @@ func SetDisplayName(
 	}
 }
 
-func getLocalPart(userID string) string {
-	if !strings.HasPrefix(userID, "@") {
-		panic(fmt.Errorf("Invalid user ID"))
+func buildMembershipEvents(
+	memberships []authtypes.Membership, db *accounts.Database,
+	newProfile authtypes.Profile, userID string, cfg *config.Dendrite,
+	queryAPI api.RoomserverQueryAPI,
+) ([]gomatrixserverlib.Event, error) {
+	evs := []gomatrixserverlib.Event{}
+
+	for _, membership := range memberships {
+		builder := gomatrixserverlib.EventBuilder{
+			Sender:   userID,
+			RoomID:   membership.RoomID,
+			Type:     "m.room.member",
+			StateKey: &userID,
+		}
+
+		content := events.MemberContent{
+			Membership: "join",
+		}
+
+		content.DisplayName = newProfile.DisplayName
+		content.AvatarURL = newProfile.AvatarURL
+
+		if err := builder.SetContent(content); err != nil {
+			return nil, err
+		}
+
+		eventsNeeded, err := gomatrixserverlib.StateNeededForEventBuilder(&builder)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ask the roomserver for information about this room
+		queryReq := api.QueryLatestEventsAndStateRequest{
+			RoomID:       membership.RoomID,
+			StateToFetch: eventsNeeded.Tuples(),
+		}
+		var queryRes api.QueryLatestEventsAndStateResponse
+		if queryErr := queryAPI.QueryLatestEventsAndState(&queryReq, &queryRes); queryErr != nil {
+			return nil, err
+		}
+
+		authEvents := gomatrixserverlib.NewAuthEvents(nil)
+
+		for i := range queryRes.StateEvents {
+			authEvents.AddEvent(&queryRes.StateEvents[i])
+		}
+
+		refs, err := eventsNeeded.AuthEventReferences(&authEvents)
+		if err != nil {
+			return nil, err
+		}
+		builder.AuthEvents = refs
+
+		eventID := fmt.Sprintf("$%s:%s", util.RandomString(16), cfg.Matrix.ServerName)
+		now := time.Now()
+		event, err := builder.Build(eventID, now, cfg.Matrix.ServerName, cfg.Matrix.KeyID, cfg.Matrix.PrivateKey)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
If we aren't then it might explain the problems that resulted in https://github.com/matrix-org/dendrite/pull/163#discussion_r128984277 . (If we aren't setting the depth and prev_events then the membership events will conflict with the existing membership events and because the depth defaults to 0 it will lose in state conflict resolution.